content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,249 @@
+
+Main critique of practical usefulness of any automated analysis approach is based on existence of *false-positives results*: issues reported as a defects, that either could not happen or are not relevant to the recipients of analysis.
+
+Traditionally, in software industry two main types of “analyzers” could be vaguely distinguished and have proved to be useful:
+ — *Style checkers* or [Linters](https://en.wikipedia.org/wiki/Lint_(software)) (i.e [CheckStyle](https://github.com/checkstyle/checkstyle), [Pylint](https://www.pylint.org/), [Golint](https://github.com/golang/lint), ets)
+ — *Bug-finding tools*, that may extend the compiler ([FindBugs](https://github.com/findbugsproject/findbugs), [Error-Prone](http://errorprone.info/), [Clang-Tidy](http://clang.llvm.org/extra/clang-tidy), [ClangThreadSafety Analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html), [Govet](https://golang.org/cmd/vet/), [CheckerFramework](https://checkerframework.org), etc)
+
+For this kind of tools, different integration points in a software development process of an organization exist, where results could be presented to the developers. They are discussed in the paper in the chronological order of BugBot team’s attempts of integrating an existing OSS static analysis tool for Java — [FindBugs](https://github.com/findbugsproject/findbugs) into the Google’s developer workflow.
+
+## History of integrating FindBugs at Google
+
+### 0. IDE/Editor
+
+Using editor as a place to show results of analysis was discarded from scratch.
+
+Cons:
+
+* many different editors/IDEs to support, as it’s before the build time
+
+* wrong time — distracting developer who may not be open to the feedback
+
+Interesting enough, [source{d}](http://sourced.tech/) as well experimented with this approach using some neural-network-based models [https://github.com/src-d/code-completion](https://github.com/src-d/code-completion) a while ago and came to the very same conclusions.
+ 
+Personally, I belive there is a huge room for improvement of code reading/writing experience though, enhancing or better augmenting developer’s abilities to both [navigate](http://kythe.io/), [understand](http://moosetechnology.org/) and [create](http://witheve.com/) programs. But indeed, existing static analysis tools which target defect detection are hardly the right tools for any of these jobs.
+
+### 2006 Bug Dashboard
+
+Next attempt included results of analysis aggregated and served from a separate web dashboard.
+
+Cons:
+
+* outside the developers’ usual workflow
+
+* distinguishing between new and existing static-analysis issues was distracting
+
+This resonates with my personal experience — in 2007, as a Build Engineer on big Java project at the first place of job, a consultancy outlet, I have experience creating internal **“Personalized Quality Dashboard”** service.
```

Comment:
"as a Build Engineer on big Java project at the first place of job, a consultancy outlet" - I can perfectly understand this after translating to Russian, but I guess this is counter-English.

Reply:
I had **an** experience of creating **the** internal